### PR TITLE
`Development`: Add endpoint for learning path competency graph

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

@JohannesWt wants a new endpoint with the following structure given the learning Path id:
```
"nodes": [
	 	{
	 		"id": 1,
	 		"label": "Competency 1",
	 		"softDueDate": "21.02.2022, 10:42"
	 		"progress": 50,
	 		"mastery": 50,
	 		"confidence": 80
	 	},
	 	{
	 		"id": 2,
	 		"label": "Competency 2",
	 		"progress": 50,
	 		"mastery": 50,
	 		"confidence": 80
	 	},
	 ],
	 "edges": [
	 	{
	 		"id": 42
	 		"source": 1,
	 		"target": 2,
	 		"type": "extends"
	 	}	
	 ]
```

### Description

Added the endpoint as requested

### Steps for Testing

No manual testing possible, since the end point is not yet used

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [LearningPathService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6626/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.learningpath/LearningPathService.html) | 99% | ✅  |
| [LearningPathResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6626/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LearningPathResource.html) | 98% | ✅  |
| [CompetencyProgressDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6626/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.competency/CompetencyProgressDTO.html) | 100% | ✅ |
| [LearningPathCompetencyGraphDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6626/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto.competency/LearningPathCompetencyGraphDTO.html) | 100% | ✅ |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/CompetencyProgressRepository.java [hunk 1]
```diff
@@ -40,6 +41,10 @@ public interface CompetencyProgressRepository extends JpaRepository<CompetencyPr
             """)
     Optional<CompetencyProgress> findByCompetencyIdAndUserId(@Param("competencyId") long competencyId, @Param("userId") long userId);
 
+    default CompetencyProgress findByCompetencyIdAndUserIdOrElseThrow(long competencyId, long userId) {
+        return findByCompetencyIdAndUserId(competencyId, userId).orElseThrow(() -> new EntityNotFoundException("CompetencyProgress"));
+    }
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/repository/CompetencyProgressRepository.java:46]: This is not the complete error message but only the name of the entity that was not found. The complete error message is built inside the exception

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java [hunk 2]
```diff
@@ -282,6 +286,15 @@ private void checkNoRelations(@NotNull Course course, @NotNull Set<LearningPathH
...
     }
 
+    public LearningPathCompetencyGraphDTO generateLearningPathCompetencyGraph(@NotNull LearningPath learningPath) {
+        Set<CompetencyProgress> progresses = competencyProgressRepository.findAllByUserIdAndLearningPathId(learningPath.getUser().getId(), learningPath.getId());
+        Set<CompetencyProgressDTO> progressDTOs = progresses.stream().map(CompetencyProgressDTO::of).collect(Collectors.toSet());
+
+        Set<CompetencyRelation> relations = competencyRelationRepository.findAllWithHeadAndTailByCourseId(learningPath.getCourse().getId());
+        Set<CompetencyRelationDTO> relationDTOs = relations.stream().map(CompetencyRelationDTO::of).collect(Collectors.toSet());
+        return new LearningPathCompetencyGraphDTO(progressDTOs, relationDTOs);
+    }
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java:296]: The processing is done inside a stream. If the set or stream are empty, then there will also not be any processing

File: src/test/java/de/tum/in/www1/artemis/competency/LearningPathIntegrationTest.java [hunk 3]
```diff
@@ -236,7 +250,7 @@ void testEnableLearningPaths() throws Exception {
     @Test
     @WithMockUser(username = INSTRUCTOR_OF_COURSE, roles = "INSTRUCTOR")
     void testEnableLearningPathsWithNoCompetencies() throws Exception {
-        var courseWithoutCompetencies = courseUtilService.createCoursesWithExercisesAndLectures(TEST_PREFIX, false, false, 0).get(0);
+        var courseWithoutCompetencies = courseUtilService.createCoursesWithExercisesAndLectures(TEST_PREFIX, false, false, 0).getFirst();
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/test/java/de/tum/in/www1/artemis/competency/LearningPathIntegrationTest.java:253]: I only refactored the access to the first element. The behavior is the exact same

File: src/main/java/de/tum/in/www1/artemis/repository/CompetencyProgressRepository.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/repository/CompetencyProgressRepository.java:46]: Yes, we customised it

**Review by JohannesWt — DISMISSED** [review]
Code LGTM 👍

**Review by JohannesWt — DISMISSED** [review]
Reapprove after adding softDueDate

File: src/main/java/de/tum/in/www1/artemis/web/rest/LearningPathResource.java [hunk 4]
```diff
@@ -278,6 +290,13 @@ public ResponseEntity<Set<CompetencyProgressForLearningPathDTO>> getCompetencyPr
         return ResponseEntity.ok(progressDTOs);
     }
 
+    private void checkLearningPathAccessElseThrow(Course course, LearningPath learningPath, User user) {
+        if (authorizationCheckService.isOnlyStudentInCourse(course, user) && !user.getId().equals(learningPath.getUser().getId())
+                && !authorizationCheckService.isAtLeastInstructorInCourse(course, user)) {
+            throw new AccessForbiddenException("You are not allowed to access another users learning path.");
```
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LearningPathResource.java:296]:
```suggestion
            throw new AccessForbiddenException("You are not allowed to access another user's learning path.");
```

**Review by pzdr7 — APPROVED** [review]
Code looks good. One small comment

**Review by JohannesWt — DISMISSED** [review]
Reapprove 👍

**Review by pzdr7 — APPROVED** [review]
Code

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by JohannesWt — APPROVED** [review]
Reapprove

**Review by pzdr7 — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java [hunk 5]
```diff
@@ -282,6 +286,15 @@ private void checkNoRelations(@NotNull Course course, @NotNull Set<LearningPathH
         }
     }
 
+    public LearningPathCompetencyGraphDTO generateLearningPathCompetencyGraph(@NotNull LearningPath learningPath) {
```
Comment (pzdr7) [inline src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java:289]: It seems the server style wants you to add a javadoc comment here:

/home/runner/work/Artemis/Artemis/src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathService.java:289:5: Missing a Javadoc comment. [MissingJavadocMethod]

**Review by pzdr7 — APPROVED** [review]

**Review by JohannesWt — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Review by JohannesWt — APPROVED** [review]
Reapprove

**Review by pzdr7 — APPROVED** [review]
Reapprove

**Pull request merged**